test/test_c10d.py

```diff
@@ -2611,6 +2635,97 @@ def forward(self, x):
         self.assertEqual(vanilla_parameter.grad, ddp_parameter.grad)
 
 
+class DistributedDataParallelDelayAllreduceTest(DistributedDataParallelTest):
+
+    @property
+    def delay_allreduce(self):
+        # all tests in ``DistributedDataParallelTest`` should pass as well
+        # under ``delay_allreduce`` mode
+        return True
+
+    @requires_nccl()
+    @skip_if_not_multigpu
+    def test_checkpoint(self):
```

Comment:
I see that we've used the sample from the github issue here to repro. Can we simplify this module to a minimal repro where we don't need the rest of the stuff in this module? Shouldn't it be possible to repro the error with a very simple module that calls `checkpoint_sequential`.

This way, it'll be easier to read and understand this unit test in the future.

Reply:
I can simplify the FFN model a bit, but we still need an additional class (`GradientStep` here) to wrap `_energy` otherwise, `checkpoint_sequential` would realize that it's the same set of parameters. Besides, I would like to make sure that it works when calling grad in forward.